### PR TITLE
fix: limit the max amount of concurrent network requests

### PR DESCRIPTION
**Pull request by zkochan**
Without limiting the amount of concurrent network requests, installation fails with too many open files error

### CODE REVIEW

File: crates/cli/src/commands/install.rs [hunk 1]
```diff
@@ -76,21 +80,26 @@ impl PackageManager {
...
-                    &self.http_client,
-                    name,
-                    version,
-                    &self.config.modules_dir,
-                )
-                .await
-                .unwrap();
-                self.install_dependencies(&dependency).await;
+            .map(|(name, version)| {
+                let semaphore_clone = semaphore.clone();
```
Comment (KSXGitHub) [inline crates/cli/src/commands/install.rs:88]: I advise `Arc::clone(&semaphore)` to avoid confusing it with regular expensive `.clone()`.

**Comment by KSXGitHub** [comment]
Is this related to the error on macOS?

File: crates/cli/src/commands/install.rs [hunk 2]
```diff
@@ -55,17 +57,19 @@ impl PackageManager {
         package
             .dependencies(self.config.auto_install_peers)
             .map(|(name, version)| async {
+                let semaphore_clone = semaphore.clone();
```
Comment (KSXGitHub) [inline crates/cli/src/commands/install.rs:60]: `semaphore` is not an `Arc` here. Cloning it would create a new `Semaphore`. Is this intentional?

**Comment by zkochan** [comment]
> Is this related to the error on macOS?

This is one of the errors on macOS. The other one is `failed to lookup address information: nodename nor servname provided, or not known`. I don't have a fix for that one yet.

Also, this PR is not ready yet. Looks like this change is causing a deadlock.

File: crates/cli/src/commands/add.rs [hunk 3]
```diff
@@ -59,13 +60,15 @@ impl PackageManager {
     /// 5. Symlink all dependencies to node_modules/.pacquet/pkg@version/node_modules
     /// 6. Update package.json
     pub async fn add(&mut self, args: &AddCommandArgs) -> Result<(), PackageManagerError> {
+        let semaphore = Arc::new(Semaphore::new(16));
```
Comment (anonrig) [inline crates/cli/src/commands/add.rs:63]: I think this should be configurable somehow

**Review by anonrig — COMMENTED** [review]
Can you add a test so that we don't regress in the future?

**Comment by KSXGitHub** [comment]
Maybe related? https://superuser.com/questions/433746/is-there-a-fix-for-the-too-many-open-files-in-system-error-on-os-x-10-7-1

**Comment by KSXGitHub** [comment]
> Looks like this change is causing a deadlock.

I don't get deadlock on my machine. What is your `package.json`?

**Comment by zkochan** [comment]
I am using [this package.json](https://raw.githubusercontent.com/pnpm/pnpm.io/main/benchmarks/fixtures/alotta-files/package.json)

**Comment by KSXGitHub** [comment]
The error is not unique to macOS, it seems.

screenshot

**Comment by zkochan** [comment]
This is a different one. I also get this one. pnpm solves this by using graceful-fs, which retries the file system operations on these type of errors.

File: crates/registry/src/package.rs [hunk 4]
```diff
@@ -30,9 +31,11 @@ impl Package {
         name: &str,
         http_client: &reqwest::Client,
         registry: &str,
+        semaphore: &Semaphore,
     ) -> Result<Self, RegistryError> {
         let url = || format!("{registry}{name}"); // TODO: use reqwest URL directly
         let network_error = |error| NetworkError { error, url: url() };
+        let _permit = semaphore.acquire().await;
```
Comment (KSXGitHub) [inline crates/registry/src/package.rs:38]: Is there a reason you gave this value a name? If you want to postpone dropping it, I recommend explicitly calling `drop` (i.e. `drop(permit)`) at the end of the scope. If you only want to `.await` the semaphore, please remove `let _permit =`.

**Comment by KSXGitHub** [comment]
> Looks like this change is causing a deadlock.

> I am using [this package.json](https://raw.githubusercontent.com/pnpm/pnpm.io/main/benchmarks/fixtures/alotta-files/package.json)

After updating the `package.json` and switching branch to `network-concurrency` (this branch), what I get is not a deadlock, but instead (it seems) am infinite loop that logs endlessly.

You may try it with `TRACE='pacquet::install,pacquet::download'` to see for yourself.

**Comment by KSXGitHub** [comment]
If I set the semaphore number to 1000, this error appears:

```text
Error:   × Main thread panicked.
  ├─▶ at crates/cli/src/package_import.rs:38:26
  ╰─▶ expected no write errors: CreateLink { from: "/home/khai/Temp/test-pacquet/pacquet-home/store/7c/a8bebf69ba6047dd7be009132ae6a30992062657634c1573fa04f35cb9de7b31d4316606d61f748300eddc0cc91220ccaa8562814d430d218a6bd0a13471e2", to:
      "/home/khai/Temp/test-pacquet/node_modules/.pacquet/dom-helpers@3.4.0/node_modules/dom-helpers/util/inDOM.js", error: Os { code: 24, kind: Uncategorized, message: "Too many open files" } }
```

But if I don't, the install process never stops.

**Comment by KSXGitHub** [comment]
I also created a more "polished" version here: https://github.com/pnpm/pacquet/commit/df491eada17de7d4022f3428ec2fac331982cb4b

Same result.

**Comment by zkochan** [comment]
I think the issue might be that pacquet doesn't support circular dependencies yet.

File: crates/registry/src/package.rs (hunk 4 above)
Comment (zkochan) [reply crates/registry/src/package.rs:38]: won't it be dropped immediately if I don't assign it to a value? As far as I understand, the permit is dropped, when it goes out of scope.

**Comment by KSXGitHub** [comment]
> I think the issue might be that pacquet doesn't support circular dependencies yet.

It would helps a lot if you can reduce your `package.json` to only circular dependencies (for testing and implementing purposes).

**Comment by zkochan** [comment]
As I have mentioned in https://github.com/pnpm/pacquet/issues/124, there are tests in pnpm related to circular dependencies. And there are small mock packages for testing: https://github.com/pnpm/registry-mock/tree/main/packages/circular-deps-1-of-2

**Pull request closed without merging**